dotCMS/src/main/java/com/dotmarketing/db/DataSourceStrategyProvider.java

```diff
@@ -0,0 +1,130 @@
+
+    @VisibleForTesting
+    DataSourceStrategyProvider(){
+        systemEnvironmentProperties = new SystemEnvironmentProperties();
+    }
+
+    private static class SingletonHelper{
+        private static final DataSourceStrategyProvider INSTANCE = new DataSourceStrategyProvider();
+    }
+
+    public static DataSourceStrategyProvider getInstance(){
+        return SingletonHelper.INSTANCE;
+    }
+
+    /**
+     * Method that loads a datasource from a custom implementation if <b>DATASOURCE_PROVIDER_STRATEGY_CLASS</b>
+     * property is defined. Otherwise, the datasource is initialized using any of these implementations (respecting order):<br>
+     * 1. A db.properties file in WEB-INF/classes implemented by {@link DBPropertiesDataSourceStrategy}<br>
+     * 2. Configuration is taken from environment variables implemented by {@link SystemEnvDataSourceStrategy}<br>
+     * 3. Getting Docker Secrets if set. Implementation: {@link DockerSecretDataSourceStrategy}<br>
+     * 4. A context.xml file in META-INF. Implementation: {@link TomcatDataSourceStrategy}
+     *
+     * @throws InstantiationException
+     * @throws IllegalAccessException
+     * @throws ClassNotFoundException
+     */
+    public DataSource get()
+            throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+
+        DataSource defaultDataSource = null;
+
+        final SystemEnvironmentProperties systemEnvironmentProperties = getSystemEnvironmentProperties();
+
+        final String providerClassName = getCustomDataSourceProvider();
+
+        if (!UtilMethods.isSet(providerClassName)) {
+            if (getDBPropertiesInstance()
+                    .existsDBPropertiesFile()) {
+                defaultDataSource = getDBPropertiesInstance()
+                        .apply();
+                Logger.info(DbConnectionFactory.class,
+                        "Datasource loaded from db.properties file");
+            } else if (systemEnvironmentProperties.getVariable("connection_db_base_url") != null) {
+                defaultDataSource = getSystemEnvDataSourceInstance()
+                        .apply();
+                Logger.info(DbConnectionFactory.class,
+                        "Datasource loaded from system environment");
+            } else if (getDockerSecretDataSourceInstance().dockerSecretPathExists()){
+                defaultDataSource = getDockerSecretDataSourceInstance()
+                        .apply();
+                Logger.info(DbConnectionFactory.class,
+                        "Datasource loaded from Docker Secret");
+            }
+
+            if (null == defaultDataSource) {
+                defaultDataSource = getTomcatDataSourceInstance()
+                        .apply();
+                Logger.info(DbConnectionFactory.class,
+                        "Datasource loaded from context.xml");
+            }
+        } else {
+            DotDataSourceStrategy customStrategy = ((Class<DotDataSourceStrategy>) Class
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Local variable 'customStrategy' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5026767)